File: modules/resin/src/com/caucho/server/resin/ResinArgs.java
/*
 * Copyright (c) 1998-2010 Caucho Technology -- all rights reserved
 *
 * This file is part of Resin(R) Open Source
 *
 * Each copy or derived work must preserve the copyright notice and this
 * notice unmodified.
 *
 * Resin Open Source is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Resin Open Source is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE, or any warranty
 * of NON-INFRINGEMENT.  See the GNU General Public License for more
 * details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Resin Open Source; if not, write to the
 *
 *   Free Software Foundation, Inc.
 *   59 Temple Place, Suite 330
 *   Boston, MA 02111-1307  USA
 *
 * @author Scott Ferguson
 */

package com.caucho.server.resin;

import java.net.Socket;
import java.util.ArrayList;
import java.util.Date;
import java.util.logging.Level;
import java.util.logging.Logger;

import com.caucho.VersionFactory;
import com.caucho.log.EnvironmentStream;
import com.caucho.log.RotateStream;
import com.caucho.server.util.JniCauchoSystem;
import com.caucho.util.L10N;
import com.caucho.vfs.Path;
import com.caucho.vfs.QJniServerSocket;
import com.caucho.vfs.QServerSocket;
import com.caucho.vfs.Vfs;
import com.caucho.vfs.WriteStream;

/**
 * The parsed Resin command-line arguments
 */
class ResinArgs
{
  private static final Logger log = Logger.getLogger(ResinArgs.class.getName());
  private static final L10N L = new L10N(ResinArgs.class);

  private final Resin _resin;
  
  private String _serverId = "";

  private Path _resinHome;
  private Path _rootDirectory;

  private String _resinConf;
  
  private DynamicServer _dynServer;

  private ArrayList<BoundPort> _boundPortList
    = new ArrayList<BoundPort>();

  private String _stage = null;
  private boolean _isDumpHeapOnExit;
  
  public ResinArgs(Resin resin, String []args)
    throws Exception
  {
    _resin = resin;
    
    preConfigureInit();
    
    parseCommandLine(args);
  }

  /**
   * Must be called after the Resin.create()
   */
  private void preConfigureInit()
  {
    String resinHome = System.getProperty("resin.home");

    if (resinHome != null)
      _resinHome = Vfs.lookup(resinHome);
    else
      _resinHome = Vfs.getPwd();

    _rootDirectory = _resinHome;

    // server.root backwards compat
    String resinRoot = System.getProperty("server.root");

    if (resinRoot != null)
      _rootDirectory = Vfs.lookup(resinRoot);

    // resin.root backwards compat
    resinRoot = System.getProperty("resin.root");

    if (resinRoot != null)
      _rootDirectory = Vfs.lookup(resinRoot);
  }
  
  public String getServerId()
  {
    return _serverId;
  }

  public Path getResinHome()
  {
    return _resinHome;
  }
  
  /**
   * Gets the root directory.
   */
  public Path getRootDirectory()
  {
    return _rootDirectory;
  }

  /**
   * The configuration file used to start the server.
   */
  public String getResinConf()
  {
    return _resinConf;
  }

  public Path getResinConfPath()
  {
    Path pwd = Vfs.lookup();

    Path resinConf = null;
    
    String resinConfFile = getResinConf();
    
    if (resinConfFile != null) {
      if (log.isLoggable(Level.FINER))
        log.finer(this + " looking for conf in " +  pwd.lookup(resinConfFile));

      resinConf = pwd.lookup(resinConfFile);
    }
    else if (pwd.lookup("conf/resin.xml").canRead())
      resinConfFile = "conf/resin.xml";
    else { // backward compat
      resinConfFile = "conf/resin.conf";
    }
    
    Path rootDirectory = getRootDirectory();

    if (resinConf == null || ! resinConf.exists()) {
      if (log.isLoggable(Level.FINER))
        log.finer(this + " looking for conf in "
                  +  rootDirectory.lookup(resinConfFile));

      resinConf = _rootDirectory.lookup(resinConfFile);
    }

    if (! resinConf.exists() && ! _resinHome.equals(_rootDirectory)) {
      if (log.isLoggable(Level.FINER))
        log.finer(this + " looking for conf in "
                  +  _resinHome.lookup(resinConfFile));

      resinConf = _resinHome.lookup(resinConfFile);
    }

    // for error messages, show path relative to rootDirectory
    if (! resinConf.exists())
      resinConf = rootDirectory.lookup(resinConfFile);

    return resinConf;
  }
  
  /**
   * Returns the bound port list.
   */
  public ArrayList<BoundPort> getBoundPortList()
  {
    return _boundPortList;
  }
  
  /**
   * Returns the stage to start with.
   */
  public String getStage()
  {
    return _stage;
  }
  
  public DynamicServer getDynamicServer()
  {
    return _dynServer;
  }
  
  public boolean isDumpHeapOnExit()
  {
    return _isDumpHeapOnExit;
  }

  private void parseCommandLine(String []argv)
    throws Exception
  {
    int len = argv.length;
    int i = 0;

    while (i < len) {
      // RandomUtil.addRandom(argv[i]);

      if (i + 1 < len
          && (argv[i].equals("-stdout")
              || argv[i].equals("--stdout"))) {
        Path path = Vfs.lookup(argv[i + 1]);

        RotateStream stream = RotateStream.create(path);
        stream.init();
        WriteStream out = stream.getStream();
        out.setDisableClose(true);

        EnvironmentStream.setStdout(out);

        i += 2;
      }
      else if (i + 1 < len
               && (argv[i].equals("-stderr")
                   || argv[i].equals("--stderr"))) {
        Path path = Vfs.lookup(argv[i + 1]);

        RotateStream stream = RotateStream.create(path);
        stream.init();
        WriteStream out = stream.getStream();
        out.setDisableClose(true);

        EnvironmentStream.setStderr(out);

        i += 2;
      }
      else if (i + 1 < len
               && (argv[i].equals("-conf")
                   || argv[i].equals("--conf"))) {
        _resinConf = argv[i + 1];
        i += 2;
      }
      else if (argv[i].equals("-log-directory")
               || argv[i].equals("--log-directory")) {
        i += 2;
      }
      else if (argv[i].equals("-config-server")
               || argv[i].equals("--config-server")) {
        i += 2;
      }
      else if (argv[i].equals("--dump-heap-on-exit")) {
        _isDumpHeapOnExit = true;

        i += 1;
      }
      else if (i + 1 < len
               && (argv[i].equals("-dynamic-server")
                   || argv[i].equals("--dynamic-server"))) {
        String []values = argv[i + 1].split(":");

        if (values.length == 3) {
          String clusterId = values[0];
          String address = values[1];
          int port = Integer.parseInt(values[2]);

          _dynServer = new DynamicServer(clusterId, address, port);
        } else {
          System.out.println("-dynamic-server requires 'cluster:address:port' at '" + argv[i + 1] + "'");

          System.exit(66);
        }

        i += 2;
      }
      else if (i + 1 < len
               && (argv[i].equals("-server")
                   || argv[i].equals("--server"))) {
        _serverId = argv[i + 1];
        i += 2;
      }
      else if (argv[i].equals("-resin-home")
               || argv[i].equals("--resin-home")) {
        _resinHome = Vfs.lookup(argv[i + 1]);

        i += 2;
      }
      else if (argv[i].equals("-root-directory")
               || argv[i].equals("--root-directory")
               || argv[i].equals("-resin-root")
               || argv[i].equals("--resin-root")
               || argv[i].equals("-server-root")
               || argv[i].equals("--server-root")) {
        _resin.setRootDirectory(_resinHome.lookup(argv[i + 1]));

        i += 2;
      }
      else if (argv[i].equals("-service")) {
        JniCauchoSystem.create().initJniBackground();
        // windows service
        i += 1;
      }
      else if (argv[i].equals("-version")
               || argv[i].equals("--version")) {
        System.out.println(VersionFactory.getFullVersion());
        System.exit(66);
      }
      else if (argv[i].equals("-watchdog-port")
               || argv[i].equals("--watchdog-port")) {
        // watchdog
        i += 2;
      }
      else if (argv[i].equals("-socketwait")
               || argv[i].equals("--socketwait")
               || argv[i].equals("-pingwait")
               || argv[i].equals("--pingwait")) {
        int socketport = Integer.parseInt(argv[i + 1]);

        Socket socket = null;
        for (int k = 0; k < 15 && socket == null; k++) {
          try {
            socket = new Socket("127.0.0.1", socketport);
          } catch (Throwable e) {
            System.out.println(new Date());
            e.printStackTrace();
          }

          if (socket == null)
            Thread.sleep(1000);
        }

        if (socket == null) {
          System.err.println("Can't connect to parent process through socket " + socketport);
          System.err.println("Resin needs to connect to its parent.");
          System.exit(0);
        }

        /*
        if (argv[i].equals("-socketwait") || argv[i].equals("--socketwait"))
          _waitIn = socket.getInputStream();
        */

        _resin.setPingSocket(socket);

        //socket.setSoTimeout(60000);

        i += 2;
      }
      else if ("-port".equals(argv[i]) || "--port".equals(argv[i])) {
        int fd = Integer.parseInt(argv[i + 1]);
        String addr = argv[i + 2];
        if ("null".equals(addr))
          addr = null;
        int port = Integer.parseInt(argv[i + 3]);

        _boundPortList.add(new BoundPort(QJniServerSocket.openJNI(fd, port),
                                         addr,
                                         port));

        i += 4;
      }
      else if ("start".equals(argv[i])
               || "restart".equals(argv[i])) {
        JniCauchoSystem.create().initJniBackground();
        i++;
      }
      else if (argv[i].equals("-verbose")
               || argv[i].equals("--verbose")) {
        i += 1;
      }
      else if (argv[i].equals("gui")) {
        i += 1;
      }
      else if (argv[i].equals("console")) {
        i += 1;
      }
      else if (argv[i].equals("-fine")
               || argv[i].equals("--fine")) {
        i += 1;
      }
      else if (argv[i].equals("-finer")
               || argv[i].equals("--finer")) {
        i += 1;
      }
      else if (argv[i].startsWith("-D")
               || argv[i].startsWith("-J")
               || argv[i].startsWith("-X")) {
        i += 1;
      }
      else if ("-stage".equals(argv[i])
               || "--stage".equals(argv[i])) {
        _resin.setStage(argv[i + 1]);
        i += 2;
      }
      else if ("-preview".equals(argv[i])
               || "--preview".equals(argv[i])) {
        _resin.setStage("preview");
        i += 1;
      }
      else if ("-debug-port".equals(argv[i])
               || "--debug-port".equals(argv[i])) {
        i += 2;
      }
      else if ("-jmx-port".equals(argv[i])
               || "--jmx-port".equals(argv[i])) {
        i += 2;
      }
      else {
        System.out.println(L.l("unknown argument '{0}'", argv[i]));
        System.out.println();
        usage();
        System.exit(66);
      }
    }
  }

  private static void usage()
  {
    System.err.println(L.l("usage: java -jar resin.jar [-options] [start | stop | restart]"));
    System.err.println(L.l(""));
    System.err.println(L.l("where options include:"));
    System.err.println(L.l("   -conf <file>          : select a configuration file"));
    System.err.println(L.l("   -log-directory <dir>  : select a logging directory"));
    System.err.println(L.l("   -resin-home <dir>     : select a resin home directory"));
    System.err.println(L.l("   -root-directory <dir> : select a root directory"));
    System.err.println(L.l("   -server <id>          : select a <server> to run"));
    System.err.println(L.l("   -watchdog-port <port> : override the watchdog-port"));
    System.err.println(L.l("   -verbose              : print verbose starting information"));
    System.err.println(L.l("   -preview              : run as a preview server"));
  }

  static class DynamicServer {
    private final String _cluster;
    private final String _address;
    private final int _port;

    DynamicServer(String cluster, String address, int port)
    {
      _cluster = cluster;
      _address = address;
      _port = port;
    }

    String getCluster()
    {
      return _cluster;
    }

    String getAddress()
    {
      return _address;
    }

    int getPort()
    {
      return _port;
    }
  }
  
  static class BoundPort {
    private QServerSocket _ss;
    private String _address;
    private int _port;

    BoundPort(QServerSocket ss, String address, int port)
    {
      if (ss == null)
        throw new NullPointerException();

      _ss = ss;
      _address = address;
      _port = port;
    }

    public QServerSocket getServerSocket()
    {
      return _ss;
    }

    public int getPort()
    {
      return _port;
    }

    public String getAddress()
    {
      return _address;
    }
  }
}

File: modules/csharp/src/common/ResinArgs.cs
/*
 * Copyright (c) 1998-2010 Caucho Technology -- all rights reserved
 *
 * This file is part of Resin(R) Open Source
 *
 * Each copy or derived work must preserve the copyright notice and this
 * notice unmodified.
 *
 * Resin Open Source is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2
 * as published by the Free Software Foundation.
 *
 * Resin Open Source is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE, or any warranty
 * of NON-INFRINGEMENT.  See the GNU General Public License for more
 * details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Resin Open Source; if not, write to the
 *
 *   Free Software Foundation, Inc.
 *   59 Temple Place, Suite 330
 *   Boston, MA 02111-1307  USA
 *
 * @author Alex Rojkov
 */
using System;
using System.Collections.Specialized;
using System.Text;

namespace Caucho
{
  public class ResinArgs
  {
    public static String USAGE = @"usage: {0} [flags] gui | console | status | start | stop | restart | kill | shutdown

    COMMANDS:
      gui - start Resin with a Graphic UI
      console - start Resin in console mode
      status - watchdog status
      start - start a Resin server
      stop - stop a Resin server
      restart - restart a Resin server
      kill - force a kill of a Resin server
      shutdown - shutdown the watchdog

    OPTIONS:
      -h                                     : this help
      -verbose                               : information on launching java
      -conf <resin.conf>                     : alternate configuration file
      -server <id>                           : select a <server> to run
      -dynamic-server <cluster:address:port> : initialize a dynamic server
      -java_home <dir>                       : sets the JAVA_HOME
      -java_exe <path>                       : path to java executable
      -classpath <dir>                       : java classpath
      -Jxxx                                  : JVM arg xxx
      -J-Dfoo=bar                            : Set JVM variable
      -Xxxx                                  : JVM -X parameter
      -resin_home <dir>                      : home of Resin
      -root-directory <dir>                  : select a root directory
      -log-directory  <dir>                  : select a logging directory
      -watchdog-port  <port>                 : override the watchdog-port
      -preview                               : run as a preview server
      -debug-port <port>                     : configure a debug port
      -jmx-port <port>                       : configure an unauthenticated jmx port
";
    public bool IsVerbose { get; private set; }
    public bool IsNoJit { get; private set; }
    public bool IsService { get; private set; }
    public bool IsInstall { get; private set; }
    public bool IsUnInstall { get; private set; }
    public bool IsHelp { get; private set; }

    //used with -install / -remove.
    //value of true indicates that the process is started
    //from resin.exe with elevated privileges
    public String User { get; private set; }
    public String Password { get; private set; }
    public String JavaExe { get; private set; }
    public String JavaHome { get; private set; }
    public String JvmArgs { get; private set; }
    public String ResinArguments { get; private set; }
    public String ClassPath { get; private set; }
    public String EnvClassPath { get; private set; }
    public String Home { get; private set; }
    public String Root { get; private set; }
    public String DisplayName { get; private set; }
    public String Command { get; private set; }
    public String ResinDataDir { get; private set; }
    public String ServiceName { get; private set; }
    public String Exe { get; private set; }
    public String Server { get; private set; }
    public String DynamicServer { get; private set; }
    public String Conf { get; private set; }
    public String Log { get; private set; }
    public String JmxPort { get; private set; }
    public String DebugPort { get; private set; }
    public String WatchDogPort { get; private set; }
    public bool IsPreview { get; private set; }

    public ResinArgs(String cmd)
    {
      int resinIdx = cmd.IndexOf("resin.exe");
      if (resinIdx == -1)
        resinIdx = cmd.IndexOf("httpd.exe");

      if (resinIdx > 0 && cmd[0] == '"')
        Exe = cmd.Substring(1, resinIdx - 1 + 9);
      else if (resinIdx > 0)
        Exe = cmd.Substring(0, resinIdx + 9);

      StringCollection arguments = new StringCollection();
      if (Exe != null)
        arguments.Add(Exe);

      StringBuilder builder = new StringBuilder();

      int startIdx = 0;
      if (resinIdx > 0)
        startIdx = resinIdx + 9;
      for (; startIdx < cmd.Length; startIdx++) {
        if (cmd[startIdx] == ' ')
          break;
      }

      bool quoted = false;
      for (int i = startIdx; i < cmd.Length; i++) {
        char c = cmd[i];
        if ('"' == c && !quoted) {
          quoted = true;
        } else if ('"' == c && quoted) {
          if (builder.Length > 0)
            arguments.Add(builder.ToString());

          builder = new StringBuilder();
          quoted = false;
        } else if (' ' == c && quoted) {
          builder.Append(c);
        } else if (' ' == c) {
          arguments.Add(builder.ToString());
          builder = new StringBuilder();

        } else if (builder != null) {
          builder.Append(c);
        } else {
          builder.Append(c);
        }
      }

      parse(arguments);
    }

    public ResinArgs(String[] args)
    {
      StringCollection arguments = new StringCollection();
      arguments.AddRange(args);

      parse(arguments);
    }

    private void parse(StringCollection arguments)
    {
      StringBuilder jvmArgs = new StringBuilder();
      StringBuilder resinArgs = new StringBuilder();
     
      int argsIdx = 1;
      while (argsIdx < arguments.Count) {
        if ("-verbose".Equals(arguments[argsIdx])) {
          IsVerbose = true;

          resinArgs.Append(' ').Append(arguments[argsIdx++]);
        } else if (arguments[argsIdx].StartsWith("-J")) {
          argsIdx++;
        } else if (arguments[argsIdx].StartsWith("-D")) {
          jvmArgs.Append(' ').Append(arguments[argsIdx]);

          argsIdx++;
        } else if (arguments[argsIdx].StartsWith("-X")) {
          jvmArgs.Append(' ').Append(arguments[argsIdx]);

          argsIdx++;
        } else if (arguments[argsIdx].StartsWith("-E")) {
          String envVar = arguments[argsIdx];
          int equalsIdx;
          if ((equalsIdx = envVar.IndexOf('=')) > -1) {
            String variable = envVar.Substring(2, equalsIdx - 2);
            String val = envVar.Substring(equalsIdx + 1);
            Environment.SetEnvironmentVariable(variable, val, EnvironmentVariableTarget.Process);
          } else {
            String variable = envVar.Substring(2);
            Environment.SetEnvironmentVariable(variable, "", EnvironmentVariableTarget.Process);
          }

          argsIdx++;
        } else if ("-user".Equals(arguments[argsIdx])) {
          User = arguments[argsIdx + 1];

          if (!User.StartsWith(".\\") && User.IndexOf('\\') < 0)
            User = ".\\" + User;

          argsIdx += 2;
        } else if ("-password".Equals(arguments[argsIdx])) {
          Password = arguments[argsIdx + 1];

          argsIdx += 2;
        } else if ("-name".Equals(arguments[argsIdx])) {
          ServiceName = arguments[argsIdx + 1];

          argsIdx += 2;
        } else if ("-display-name".Equals(arguments[argsIdx])) {
          DisplayName = arguments[argsIdx + 1];

          argsIdx += 2;
        } else if ("-install".Equals(arguments[argsIdx])) {
          IsInstall = true;
          IsService = false;

          argsIdx++;
        } else if ("-install-as".Equals(arguments[argsIdx]) ||
                   "-install_as".Equals(arguments[argsIdx])) {
          IsInstall = true;
          IsService = false;

          ServiceName = arguments[argsIdx + 1];
          DisplayName = arguments[argsIdx + 1];

          argsIdx += 2;
        } else if ("-remove".Equals(arguments[argsIdx])) {
          IsUnInstall = true;
          IsService = false;

          argsIdx++;
        } else if ("-remove-as".Equals(arguments[argsIdx]) ||
                   "-remove_as".Equals(arguments[argsIdx])) {
          IsUnInstall = true;
          IsService = false;

          ServiceName = arguments[argsIdx + 1];

          argsIdx += 2;
        } else if ("-java_home".Equals(arguments[argsIdx]) ||
                  "-java-home".Equals(arguments[argsIdx])) {
          JavaHome = arguments[argsIdx + 1];

          argsIdx += 2;
        } else if ("-java_exe".Equals(arguments[argsIdx]) ||
                   "-java-exe".Equals(arguments[argsIdx])) {
          JavaExe = arguments[argsIdx + 1];

          argsIdx += 2;
        } else if ("-msjava".Equals(arguments[argsIdx])) {
          //msJava = true; XXX no longer supported

          argsIdx++;
        } else if ("-resin_home".Equals(arguments[argsIdx]) ||
                   "-resin-home".Equals(arguments[argsIdx]) ||
                   "--resin-home".Equals(arguments[argsIdx])) {
          Home = arguments[argsIdx + 1];

          argsIdx += 2;
        } else if ("-server-root".Equals(arguments[argsIdx]) ||
                   "-server_root".Equals(arguments[argsIdx]) ||
                   "--root-directory".Equals(arguments[argsIdx]) ||
                   "-root-directory".Equals(arguments[argsIdx])) {
          Root = arguments[argsIdx + 1];

          argsIdx += 2;
        } else if ("-classpath".Equals(arguments[argsIdx]) ||
                   "-cp".Equals(arguments[argsIdx])) {
          ClassPath += arguments[argsIdx + 1];

          argsIdx += 2;
        } else if ("-env-classpath".Equals(arguments[argsIdx])) {
          EnvClassPath = arguments[argsIdx + 1];

          argsIdx += 2;
        } else if ("-stdout".Equals(arguments[argsIdx])) {
          //stdOutFile = args[argsIdx + 1]; XXX not supported

          argsIdx += 2;
        } else if ("-stderr".Equals(arguments[argsIdx])) {
          //stdErrFile = args[argsIdx + 1]; XXX not supported

          argsIdx += 2;
        } else if ("-jvm-log".Equals(arguments[argsIdx])) {
          //jvmFile = args[argsIdx + 1]; XXX not supported

          argsIdx += 2;
        } else if ("-main".Equals(arguments[argsIdx])) {
          //main = args[argsIdx + 1]; XXX not supported - was used with jview

          argsIdx += 2;
        } else if ("-help".Equals(arguments[argsIdx]) ||
                  "-h".Equals(arguments[argsIdx])) {
          IsHelp = true;

          argsIdx++;
        } else if ("-service".Equals(arguments[argsIdx])) {
          IsService = true;

          argsIdx++;
        } else if ("-console".Equals(arguments[argsIdx])) {
          //make -console be a command
          Command = "console";

          argsIdx++;
        } else if ("-nojit".Equals(arguments[argsIdx])) {
          IsNoJit = true;

          argsIdx++;
        } else if ("-standalone".Equals(arguments[argsIdx])) {
          argsIdx++;
        } else if ("-server".Equals(arguments[argsIdx]) ||
                   "--server".Equals(arguments[argsIdx])) {
          Server = arguments[argsIdx + 1];

          argsIdx += 2;
        } else if ("-dynamic-server".Equals(arguments[argsIdx]) ||
                   "--dynamic-server".Equals(arguments[argsIdx])) {
          DynamicServer = arguments[argsIdx + 1];

          argsIdx += 2;
        } else if ("-conf".Equals(arguments[argsIdx]) ||
                   "--conf".Equals(arguments[argsIdx])) {
          Conf = arguments[argsIdx + 1];

          resinArgs.Append("-conf \"").Append(Conf).Append("\" ");

          argsIdx += 2;
        } else if ("-log-directory".Equals(arguments[argsIdx]) ||
                   "--log-directory".Equals(arguments[argsIdx])) {
          Log = arguments[argsIdx + 1];

          resinArgs.Append("-log-directory \"").Append(Log).Append("\" ");

          argsIdx += 2;
        } else if ("-jmx-port".Equals(arguments[argsIdx]) ||
                   "--jmx-port".Equals(arguments[argsIdx])) {
          JmxPort = arguments[argsIdx + 1];

          resinArgs.Append("-jmx-port ").Append(JmxPort).Append(' ');

          argsIdx += 2;
        } else if ("-debug-port".Equals(arguments[argsIdx]) ||
                   "--debug-port".Equals(arguments[argsIdx])) {
          DebugPort = arguments[argsIdx + 1];

          resinArgs.Append("-debug-port ").Append(DebugPort).Append(' ');

          argsIdx += 2;
        } else if ("-watchdog-port".Equals(arguments[argsIdx]) ||
                   "--watchdog-port".Equals(arguments[argsIdx])) {
          WatchDogPort = arguments[argsIdx + 1];

          resinArgs.Append("-watchdog-port ").Append(WatchDogPort).Append(' ');

          argsIdx += 2;
        } else if ("-e".Equals(arguments[argsIdx]) ||
                  "-compile".Equals(arguments[argsIdx])) {
          argsIdx++;
        } else if ("-preview".Equals(arguments[argsIdx]) ||
                  "--preview".Equals(arguments[argsIdx])) {
          IsPreview = true;
          argsIdx++;
        } else if ("gui".Equals(arguments[argsIdx])) {
          Command = arguments[argsIdx];

          argsIdx++;
        } else if ("console".Equals(arguments[argsIdx]) ||
                   "status".Equals(arguments[argsIdx]) ||
                   "start".Equals(arguments[argsIdx]) ||
                   "gui".Equals(arguments[argsIdx]) ||
                   "stop".Equals(arguments[argsIdx]) ||
                   "restart".Equals(arguments[argsIdx]) ||
                   "kill".Equals(arguments[argsIdx]) ||
                   "shutdown".Equals(arguments[argsIdx])) {
          Command = arguments[argsIdx];

          argsIdx++;
        } else {
          resinArgs.Append(' ').Append(arguments[argsIdx++]).Append(' ');
        }
      }

      if (ServiceName == null || "".Equals(ServiceName))
        ServiceName = "Resin";

      JvmArgs = jvmArgs.ToString();
      ResinArguments = resinArgs.ToString();
    }

    public bool IsValid()
    {
      return Command != null || IsInstall || IsUnInstall || IsService;
    }

    public bool IsServiceCommand()
    {
      return IsInstall || IsUnInstall;
    }
  }
}